# 10.5 Matriser

Her ser vi på **produktet** av matriser 

Produktet en radvektor og kolonnevektor med samme lengde er gitt ved å legge kolonnevektoren ned slik at den blir til en radvektor og så ta prikkproduktet av de to radvektorene. 

## Eksempel

$$
\left[
\begin{array}{cccc}
    1 & 2 & 3 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2\\
3\\
1\\
-1
\end{array}
\right]
=
\left[
\begin{array}{c}
    1 & 2 & 3 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
3&
1&
-1
\end{array}
\right]
=
2 + 6 + 3 - 2 = 9
$$

**Merk:** rad- og kolonnevektor må komme i denne rekkefølgen.

---

Denne regelen kan utvides til tilfellet der vi har mere enn en rad.

## Eksempel 

Her multipliserer vi en $2 \times 4$-matrise med en $4 \times 1$-matrise

$$
\left[
\begin{array}{cccc}
    1 & 2 & 3 & 2 \\
    2 & -1 & 2 & 1
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2\\
3\\
1\\
-1
\end{array}
\right]
=
\left[
\begin{array}{c}
\left[
\begin{array}{c}
    1 & 2 & 3 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
3&
1&
-1
\end{array}
\right]
\\
\left[
\begin{array}{c}
    2 & -1 & 2 & 1 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
3&
1&
-1
\end{array}
\right]
\end{array}
\right]
=
\left[
\begin{array}{c}
2 + 6 + 3 - 2 \\
4 - 3 + 2 - 1 
\end{array}
\right]=
\left[
\begin{array}{c}
9\\
2
\end{array}
\right]
$$

[Her er et eksempel fra støkiometri.](https://colab.research.google.com/github/mbr085/V25MAT102/blob/main/notebooks/uke7/stoichiometri.ipynb)

---

Vi kan også utvide regelen til tilfellet der vi har mere enn en kolonne.

## Eksempel

Her multipliserer vi en $1 \times 2$-matrise med en $2 \times 2$-matrise:

$$
\left[
\begin{array}{cc}
    1 & 2
\end{array}
\right]
\cdot 
\left[
\begin{array}{cc}
2&
3\\
1&
-1
\end{array}
\right]
=
\left[
\begin{array}{c}
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
1
\end{array}
\right]
&
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
3&
-1
\end{array}
\right]
\end{array}
\right]
=
\left[
\begin{array}{c}
2 + 2
&
3 - 2
\end{array}
\right]
=
\left[
\begin{array}{c}
4
&
1
\end{array}
\right]
$$

---

Slår vi de to utvidelsene sammen kan vi multiplisere en $m \times n$-matrise med en $n \times p$-matrise. Formelen står i læreboken.

$$
\left[
\begin{array}{cc}
    1 & 2 \\
    3 & 4
\end{array}
\right]
\cdot 
\left[
\begin{array}{cc}
2&
3\\
1&
-1
\end{array}
\right]
=
\left[
\begin{array}{c}
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
1
\end{array}
\right]
&
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
3&
-1
\end{array}
\right]
\\
\left[
\begin{array}{c}
    3 & 4 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
1
\end{array}
\right]
&
\left[
\begin{array}{c}
    3 & 4 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
3&
-1
\end{array}
\right]
\end{array}
\right]
=
\left[
\begin{array}{c}
2 + 2
&
3 - 2
\\
6 + 4
&
9 - 4
\end{array}
\right]
=
\left[
\begin{array}{c}
4
&
1\\
10 & 5
\end{array}
\right]
$$

## Eksempel

La oss se på et eksempel på næring. La oss si at vi har følgende tabell for innhold av prosent næringsinnhold i fire typer mat:

\begin{array}{l|cccc}
 & Brød & Melk & Ost & Smør\\
\hline 
Protein & 7,5 & 3,4 & 27,5 & 0,5 \\
Fett & 1,2 & 3,8 & 28,6 & 81,3\\ 
Karbohydrat & 54,5 & 4,8 & 2,0 & 0,0
\end{array}

Spørsmålet "hvis du til frokost fikk $100g$ brød, $200g$ melk, $20g$ ost og $10g$ smør, hvor mye fikk du da av de forskjellige næringsstoffene?"
kan besvares ved hjelp av multiplikasjon av matriser:

$$
\left[
\begin{array}{cccc}
    7.5 & 3.4 & 27.5 & 0.5 \\
    1.2 & 3.8 & 28.6 & 81.3 \\
    54.5 & 4.8 & 2.0 & 0
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
100\\
200\\
20\\
10
\end{array}
\right]
$$

## Oppgave

**Forklar** til de rundt deg hvordan du kan lese av fra dette matriseproduktet hvor mye du fikk av de forskjellige næringsstoffene. **Regn det ut** når du har forstått hvordan det fungerer.

Her brukes numpy til å beregne matriseproduktet. Hvorfor deler vi på $100$?

In [1]:
import numpy as np

A = np.array([
    [7.5, 3.4, 27.5, 0.5],
    [1.2, 3.8, 28.6, 81.3],
    [54.5, 4.8, 2.0, 0]
])

v = np.array([
    [100],
    [200],
    [20],
    [10]
])

A @ v / 100

array([[19.85],
       [22.65],
       [64.5 ]])

---

La oss regne ut hvor mye energi måltidet ga. Følgende er funnet på [en nettside til BBC](https://www.bbc.co.uk/bitesize/guides/zqj66yc/revision/4)

Karbohydrater, proteiner og fett gir energien som kroppen får fra mat.  

Hvert næringsstoff gir en ulik mengde energi, målt i kilokalorier:  

- 1 gram protein gir 4 kilokalorier.  
- 1 gram fett gir 9 kilokalorier.
- 1 gram karbohydrat gir 3,75 kilokalorier.  

Vi fant ut at måltiden er fordelt på $19.85g$ protein, $22.65g$ fett og $64.5g$ karbohydrat. Dette kan sammenfattes i vektoren
$$
\left[
\begin{array}{c}
  19.85\\
  22.65\\
  64.5
\end{array}
\right]
$$


**Forklar** hvorfor matriseproduktet

$$
\left[
\begin{array}{ccc}
  4 & 9 & 3.75
\end{array}
\right]
\cdot
\left[
\begin{array}{c}
  19.85\\
  22.65\\
  64.5
\end{array}
\right]
$$


gir antall kilokalorier i måltidet. **Regn det ut**.

**Forklar til de rundt deg hva følgende matriseprodukt regner ut**

$$
\left[
\begin{array}{ccc}
  4 & 9 & 3.75
\end{array}
\right]
\cdot
\left[
\begin{array}{cccc}
    7.5 & 3.4 & 27.5 & 0.5 \\
    1.2 & 3.8 & 28.6 & 81.3 \\
    54.5 & 4.8 & 2.0 & 0
\end{array}
\right]
$$

**Forklar til de rundt deg hvorfor følgende likhet er sann.** Hva regnes ut? 

$$
\left(
\left[
\begin{array}{ccc}
  4 & 9 & 3.75
\end{array}
\right]
\cdot
\left[
\begin{array}{cccc}
    7.5 & 3.4 & 27.5 & 0.5 \\
    1.2 & 3.8 & 28.6 & 81.3 \\
    54.5 & 4.8 & 2.0 & 0
\end{array}
\right]
\right)
\cdot 
\left[
\begin{array}{c}
1\\
2\\
0.2\\
0.1
\end{array}
\right]
=
\left[
\begin{array}{ccc}
  4 & 9 & 3.75
\end{array}
\right]
\cdot
\left(\left[
\begin{array}{cccc}
    7.5 & 3.4 & 27.5 & 0.5 \\
    1.2 & 3.8 & 28.6 & 81.3 \\
    54.5 & 4.8 & 2.0 & 0
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
1\\
2\\
0.2\\
0.1
\end{array}
\right]\right)
$$

## Regneregler for matriser

Vi kan regne med matriser nesten som med vanlige tall. (Se Teorem 1 i Kapittel 10.5 i MIP.) Ved matrisemultiplikasjon må vi må passe på at vi ganger sammen matriser av passende format. Ellers er en største forskjellen er at for to tall $a$ og $b$ er $ab = ba$. Denne likheten holder ikke for matriser $A$ og $B$. For det første kan formatet passe med den ene multiplikasjonen og ikke med den andre. For det andre, selv om både $AB$ og $BA$ passer med formatet kan matrisene $AB$ og $BA$ være forskjellige. Dessuten kan $AB$ være null selv om verken $A$ eller $B$ er null.

## Eksempel

Vi har regnet ut at 

$$
\left[
\begin{array}{cc}
    1 & 2 \\
    3 & 4
\end{array}
\right]
\cdot 
\left[
\begin{array}{cc}
2&
3\\
1&
-1
\end{array}
\right]
=
\left[
\begin{array}{c}
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
1
\end{array}
\right]
&
\left[
\begin{array}{c}
    1 & 2 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
3&
-1
\end{array}
\right]
\\
\left[
\begin{array}{c}
    3 & 4 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
1
\end{array}
\right]
&
\left[
\begin{array}{c}
    3 & 4 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
3&
-1
\end{array}
\right]
\end{array}
\right]
=
\left[
\begin{array}{c}
2 + 2
&
3 - 2
\\
6 + 4
&
9 - 4
\end{array}
\right]
=
\left[
\begin{array}{c}
4
&
1\\
10 & 5
\end{array}
\right]
$$

Bytter vi om på rekkefølgen får vi

$$
\left[
\begin{array}{cc}
2&
3\\
1&
-1
\end{array}
\right]
\cdot
\left[
\begin{array}{cc}
    1 & 2 \\
    3 & 4
\end{array}
\right]
=
\left[
\begin{array}{c}
\left[
\begin{array}{c}
    2 & 3 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
1&
3
\end{array}
\right]
&
\left[
\begin{array}{c}
    2 & 3 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
4
\end{array}
\right]
\\
\left[
\begin{array}{c}
    1 & -1 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
1&
3
\end{array}
\right]
&
\left[
\begin{array}{c}
    1 & -1 
\end{array}
\right]
\cdot 
\left[
\begin{array}{c}
2&
4
\end{array}
\right]
\end{array}
\right]
=
\left[
\begin{array}{c}
2 + 9
&
4 + 12
\\
1 -3
&
2-4
\end{array}
\right]
=
\left[
\begin{array}{c}
11
&
16\\
-2 & -2
\end{array}
\right]
$$

## Eksempel

La 
$$
A = \left[\begin{array}{cc}0&1\\0&0\end{array}\right]
$$
Da er $AA = 0$.

## Quiz

Her er et [spørsmål fra eksamen i fjor](https://mitt.uib.no/courses/50911/quizzes/48452)